src/app.rs

```diff
@@ -71,116 +62,99 @@ use crate::{
 /// Where to go from here: Please see [`Router`](struct.Router.html) and [`Endpoint`](trait.Endpoint.html)
 /// for further examples.
 ///
-pub struct App<Data> {
-    data: Data,
-    router: Router<Data>,
-    default_handler: EndpointData<Data>,
+/// A core type for routing.
+///
+/// The `App` type can be used to set up routes and resources, and to apply middleware.
+pub struct App<AppData> {
+    router: Router<AppData>,
+    middleware: Vec<Arc<dyn Middleware<AppData>>>,
+    data: AppData,
 }
 
-impl<Data: Clone + Send + Sync + 'static> App<Data> {
-    /// Set up a new app with some initial `data`.
-    pub fn new(data: Data) -> App<Data> {
-        let logger = RootLogger::new();
-        let mut app = App {
-            data,
+impl<AppData: Clone + Send + Sync + 'static> App<AppData> {
+    /// Create an empty `App`, with no initial middleware or configuration.
+    pub fn new(data: AppData) -> App<AppData> {
+        App {
             router: Router::new(),
-            default_handler: EndpointData {
-                endpoint: BoxedEndpoint::new(async || http::status::StatusCode::NOT_FOUND),
-                store: Store::new(),
-            },
-        };
-
-        // Add RootLogger as a default middleware
-        app.middleware(logger);
-        app.setup_configuration();
-
-        app
-    }
-
-    // Add default configuration
-    fn setup_configuration(&mut self) {
-        let config = Configuration::build().finalize();
-        self.config(config);
-    }
-
-    /// Get the top-level router.
-    pub fn router(&mut self) -> &mut Router<Data> {
-        &mut self.router
-    }
-
-    /// Add a new resource at `path`.
-    /// See [Router.at](struct.Router.html#method.at) for details.
-    pub fn at<'a>(&'a mut self, path: &'a str) -> Resource<'a, Data> {
-        self.router.at(path)
-    }
-
-    /// Set the default handler for the app, a fallback function when there is no match to the route requested
-    pub fn default_handler<T: Endpoint<Data, U>, U>(
-        &mut self,
-        handler: T,
-    ) -> &mut EndpointData<Data> {
-        let endpoint = EndpointData {
-            endpoint: BoxedEndpoint::new(handler),
-            store: self.router.store_base.clone(),
-        };
-        self.default_handler = endpoint;
-        &mut self.default_handler
-    }
-
-    /// Apply `middleware` to the whole app. Note that the order of nesting subrouters and applying
-    /// middleware matters; see `Router` for details.
-    pub fn middleware(&mut self, middleware: impl Middleware<Data> + 'static) -> &mut Self {
-        self.router.middleware(middleware);
-        self
-    }
```

Comment:
Silly q: I was poking around trying to build an app with middleware against this branch (I'm excited for these changes!), but noticed that `app.middleware()` no longer works – was this method removed intentionally / is it coming back in a different form? (Or have I just jumped the gun entirely? 😅)

Reply:
@chrisdickinson Good question! -- I hadn't realized this was missing until now. Middleware is staying as a concept, so it should be coming back in some form. Probably more on this this week!